prov/shm/src/smr_ep.c

```diff
@@ -460,6 +449,9 @@ int smr_endpoint(struct fid_domain *domain, struct fi_info *info,
 	smr_init_queue(&ep->trecv_queue, smr_match_tagged);
 	smr_init_queue(&ep->unexp_queue, smr_match_unexp);
 
+	ep->util_ep.min_multi_recv_size = ep->util_ep.rx_op_flags &
+			FI_MULTI_RECV ? SMR_INJECT_SIZE : 0;
+
```

Comment:
This doesn't look right.  op_flags are default flags to apply to a receive operation.  That should be separate from the minimum multi-recv size.  FI_MULTI_RECV is a secondary capability, so it's always safe to enable for an application.  An app can change the default op_flags using fi_control() or by calling fi_recvmsg().

Reply:
Gotcha, that makes sense to me. I was mirroring the initialization in the rxm provider. Should that be following the same idea? (rxm_ep.c:1668)

Reply:
@a-ilango - rxm looks wrong to me in this same area

Reply:
I can go ahead and include that in this PR. Or do you want me to leave it to you since I'll be removing those rxm changes @a-ilango ?

Reply:
The inject size is same as that of bounce buffer which is used for eager messages in RxM provider. Keeping min_multi_recv_size same as this would ensure that we always have space in the user recv buffer when we copy data in eager messaging.

Reply:
Sorry, I misread the question. :) I'll prepare a separate PR for rxm to fix setting of min_multi_recv_size.

Reply:
The minimum multi-recv size is a user controlled value.  It is independent of any op flags being specified.  It applies to the user's posted receive buffer only and when that buffer should be freed. When the amount of free space in the user buffer drops below this value, it should be returned to the app.

An app can always send a message that is larger than the space remaining in the buffer.  If that happens there are a couple of options.  One is to truncate the message and report that to the user.  The second is to report that the multi-recv buffer has been released, and copy the received message into the next posted buffer (assuming that it is large enough to hold it).

There really should not be any association between min multi-recv size and op flags or the inject size.

Reply:
So are you saying we should always initialize to 0?

Reply:
> The second is to report that the multi-recv buffer has been released, and copy the received message into the next posted buffer (assuming that it is large enough to hold it).

iiuc, rxm returns a truncation error when there isn't enough space in the user buffer. But the above idea seems to be better in terms of error handling. Should the next buffer also be posted with FI_MULTI_RECV flag for this behavior to be valid?